test/e2e_federation/util.go

```diff
@@ -172,33 +226,70 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
+	err := wait.PollImmediate(framework.Poll, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
+		var err error
+		cSvc, err = cluster.Clientset.Services(namespace).Get(serviceName, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			// Get failed with an error, try again.
+			framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v. Trying again in %s", serviceName, namespace, name, err, framework.Poll)
+			return false, nil
+		} else if errors.IsNotFound(err) {
+			cSvc = nil
+			By(fmt.Sprintf("Service %q in namespace %q in cluster %q not found", serviceName, namespace, name))
+			return true, err
+		}
+		By(fmt.Sprintf("Service %q in namespace %q in cluster %q found", serviceName, namespace, name))
+		return true, nil
+	})
+
+	return cSvc, err
+}
+
+func getServiceShardWithRetry(namespace string, serviceName string, cluster *fedframework.Cluster) (*v1.Service, error) {
+	name := cluster.Name
+	var cSvc *v1.Service
+	err := wait.PollImmediate(framework.Poll, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
+		var err error
+		cSvc, err = cluster.Clientset.Services(namespace).Get(serviceName, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			// Get failed with an error, try again.
+			framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v. Trying again in %s", serviceName, namespace, name, err, framework.Poll)
+			return false, nil
+		} else if errors.IsNotFound(err) {
+			cSvc = nil
+			framework.Logf("Service %q in namespace %q in cluster %q not found. Trying again in %s", serviceName, namespace, name, framework.Poll)
+			return false, nil
+		}
+		By(fmt.Sprintf("Service %q in namespace %q in cluster %q found", serviceName, namespace, name))
+		return true, nil
+	})
+
+	return cSvc, err
+}
+
 func cleanupServiceShardsAndProviderResources(namespace string, service *v1.Service, clusters fedframework.ClusterSlice) {
 	framework.Logf("Deleting service %q in %d clusters", service.Name, len(clusters))
 	for _, c := range clusters {
 		name := c.Name
-		var cSvc *v1.Service
-
-		err := wait.PollImmediate(framework.Poll, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
-			var err error
-			cSvc, err = c.Clientset.Services(namespace).Get(service.Name, metav1.GetOptions{})
-			if err != nil && !errors.IsNotFound(err) {
-				// Get failed with an error, try again.
-				framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v. Trying again in %s", service.Name, namespace, name, err, framework.Poll)
-				return false, nil
-			} else if errors.IsNotFound(err) {
-				cSvc = nil
-				By(fmt.Sprintf("Service %q in namespace %q in cluster %q not found", service.Name, namespace, name))
-				return true, err
-			}
-			By(fmt.Sprintf("Service %q in namespace %q in cluster %q found", service.Name, namespace, name))
-			return true, err
-		})
+		cSvc, err := getServiceShard(namespace, service.Name, c)
```

Comment:
@cedriclam  There was no need of pulling out code and creating `getServiceShard` function. Can you please revert it. I would apply the LGTM after this change. Thanks !